File: app/src/main/java/com/example/fruitsdiary/usecase/addeditentry/AddEditEntryIntent.java
package com.example.fruitsdiary.usecase.addeditentry;

import android.content.Context;
import android.content.Intent;
import android.support.annotation.IntDef;
import android.support.annotation.Nullable;

import com.example.fruitsdiary.model.Entry;

import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;

import static com.example.fruitsdiary.usecase.addeditentry.AddEditEntryIntent.EntryState.CREATE;
import static com.example.fruitsdiary.usecase.addeditentry.AddEditEntryIntent.EntryState.VIEW;
import static com.example.fruitsdiary.usecase.addeditentry.AddEditEntryIntent.EntryState.EDIT;

public class AddEditEntryIntent extends Intent {

    @IntDef({CREATE, VIEW, EDIT})
    @Retention(RetentionPolicy.SOURCE)
    public @interface EntryState {
        int CREATE = 0;
        int VIEW = 1;
        int EDIT = 2;
    }

    private static final String ARGS_ENTRY = "ARGS_ENTRY";
    private static final String ARGS_ENTRY_STATE = "ARGS_ENTRY_STATE";

    public AddEditEntryIntent(Context context) {
        super(context, AddEditEntryActivity.class);
        putExtra(ARGS_ENTRY_STATE, CREATE);
    }

    public AddEditEntryIntent(Context context, @Nullable Entry entry) {
        super(context, AddEditEntryActivity.class);
        putExtra(ARGS_ENTRY_STATE, VIEW);
        putExtra(ARGS_ENTRY, entry);
    }

    public AddEditEntryIntent(Intent intent) {
        super(intent);
    }

    public Entry getEntry() {
        return getParcelableExtra(ARGS_ENTRY);
    }

    @EntryState
    public int getEntryState(){
        return getIntExtra(ARGS_ENTRY_STATE, VIEW);
    }
}

File: app/src/main/java/com/example/fruitsdiary/usecase/addeditentry/AddEditEntryIntent.kt
package com.example.fruitsdiary.usecase.addeditentry

import android.content.Context
import android.content.Intent
import com.example.fruitsdiary.model.Entry
import com.example.fruitsdiary.usecase.addeditentry.AddEditEntryIntent.EntryState.Companion.CREATE
import com.example.fruitsdiary.usecase.addeditentry.AddEditEntryIntent.EntryState.Companion.VIEW

class AddEditEntryIntent : Intent {

    annotation class EntryState {
        companion object {
            const val CREATE = 0
            const val VIEW = 1
            const val EDIT = 2
        }
    }

    constructor(context: Context?) : super(context, AddEditEntryActivity::class.java) {
        putExtra(ARGS_ENTRY_STATE, CREATE)
    }

    constructor(context: Context?, entry: Entry?) : super(context, AddEditEntryActivity::class.java) {
        putExtra(ARGS_ENTRY_STATE, VIEW)
        putExtra(ARGS_ENTRY, entry)
    }

    constructor(intent: Intent) : super(intent)

    fun getEntry(): Entry = getParcelableExtra(ARGS_ENTRY)

    @EntryState
    fun getEntryState(): Int = getIntExtra(ARGS_ENTRY_STATE, VIEW)

    companion object {
        private const val ARGS_ENTRY = "ARGS_ENTRY"
        private const val ARGS_ENTRY_STATE = "ARGS_ENTRY_STATE"
    }
}
